src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php

```diff
@@ -76,11 +91,198 @@ public function indexAction(Request $request, MerchandiseReturnFilters $filters)
+        $formHandler = $this->get('prestashop.core.form.identifiable_object.handler.merchandise_return_form_handler');
+        $gridFactory = $this->get('prestashop.core.grid.factory.merchandise_return_products');
+
+        try {
+            /** @var EditableMerchandiseReturn $editableMerchandiseReturn */
+            $editableMerchandiseReturn = $this->getQueryBus()->handle(
+                new GetMerchandiseReturnForEditing(
+                    $merchandiseReturnId
+                )
+            );
+
+            $form = $formBuilder->getFormFor($merchandiseReturnId);
+            $form->handleRequest($request);
+
+            $result = $formHandler->handleFor($merchandiseReturnId, $form);
+
+            if ($result->getIdentifiableObjectId()) {
+                $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));
+
+                return $this->redirectToRoute('admin_merchandise_returns_index');
+            }
+        } catch (Exception $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages()));
+        }
+
+        if (!isset($form)) {
+            return $this->redirectToRoute('admin_merchandise_returns_index');
+        }
+
+        return $this->render('@PrestaShop/Admin/Sell/CustomerService/MerchandiseReturn/edit.html.twig', [
+            'help_link' => $this->generateSidebarLink($request->attributes->get('_legacy_controller')),
+            'enableSidebar' => true,
+            'layoutTitle' => sprintf($this->trans('Return Merchandise Authorization (RMA) ', 'Admin.Actions')),
+            'merchandiseReturnForm' => $form->createView(),
+            'editableMerchandiseReturn' => $editableMerchandiseReturn,
+            'merchandiseReturnsProductsGrid' => $this->presentGrid($gridFactory->getGrid($filters)),
+        ]);
+    }
+
+    /**
+     * @AdminSecurity("is_granted('delete', request.get('_legacy_controller'))", redirectRoute="admin_merchandise_returns_index")
+     *
+     * @param Request $request
+     * @param int $merchandiseReturnId
+     * @param int $merchandiseReturnDetailId
+     * @param int $customizationId
+     *
+     * @return RedirectResponse
+     */
+    public function deleteProductAction(Request $request, int $merchandiseReturnId, int $merchandiseReturnDetailId, int $customizationId): RedirectResponse
+    {
+        try {
+            $this->getCommandBus()->handle(
+                new DeleteProductFromMerchandiseReturnCommand($merchandiseReturnId, $merchandiseReturnDetailId, $customizationId)
+            );
+
+            $this->addFlash(
+                'success',
+                $this->trans('Successful deletion.', 'Admin.Notifications.Success')
+            );
+        } catch (Exception $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages($e)));
+        }
+
+        return $request->query->has('redirectUrl') ?
```

Comment:
Can you explain what happens here ? because I have never seen this in other "delete" actions
So it should be one of the 3:
- maybe it's not needed to be done, it's already being done by some other logic like a Symfony listener
- maybe it needs to be done 😄 you have found something, but we should update _ALL_ delete actions we have to follow it
- maybe it's only for merchandise return

Reply:
When I looked for example first thing I clicked on is AddressController, turns out it's the only controller that does things this way.

   ```
     return $request->query->has('redirectUrl') ?
            $this->redirect($request->query->get('redirectUrl')) :
            $this->redirectToRoute('admin_addresses_index');
```

Don't really know why it was done like this in Address controller. At that moment I  copied it I thought it's done like this everywhere else and it's a simple check that I should keep. But it works fine without so I can remove the if.

Reply:
I think there must be a controller that "sends people to address controller then expects them to come back after they have successfully created/updated the address" (probably order page).

1. user clicks on button "edit address"
2. user is redirected on address form
3. user does the changes
4. user is redirected to first page

This should be done everywhere or nowhere. We can remove it for now and we need to explore if this is a feature.

@MatShir do you know about this ?